walk/walk.go

```diff
@@ -162,32 +162,38 @@ func Walk(c *config.Config, cexts []config.Configurer, dirs []string, mode Mode,
 			}
 		}
 
-		shouldUpdate := shouldUpdate(rel, mode, updateParent, updateRels)
+		shouldUpdate := updateRels.shouldUpdate(rel, updateParent)
 		for _, sub := range subdirs {
-			if subRel := path.Join(rel, sub); shouldVisit(subRel, mode, shouldUpdate, updateRels) {
+			if subRel := path.Join(rel, sub); updateRels.shouldVisit(subRel, shouldUpdate) {
 				visit(c, filepath.Join(dir, sub), subRel, shouldUpdate)
 			}
 		}
 
 		update := !haveError && !wc.ignore && shouldUpdate
-		if shouldCall(rel, mode, updateParent, updateRels) {
+		if updateRels.shouldCall(rel, updateParent) {
 			genFiles := findGenFiles(wc, f)
 			wf(dir, rel, c, update, f, subdirs, regularFiles, genFiles)
 		}
 	}
 	visit(c, c.RepoRoot, "", false)
 }
 
-// buildUpdateRelMap builds a table of prefixes, used to determine which
+// An UpdateFilter tracks which directories need to be updated
+type UpdateFilter struct {
```

Comment:
Since we are moving these types to a public package, could you add a comment marking them as internal? Alternatively, could we carve out a new `internal` subpackage and move them there?

Reply:
For now I can actually make this struct private which I've done in this PR.

But for https://github.com/bazelbuild/bazel-gazelle/pull/1880 this logic of which directories are updated is needed when deciding what data to load vs re-compute.

Reply:
I've reverted back to using `UpdateFilter` (making it go-lang public) so it can be used elsewhere in the next PR. I've added the INTERNAL comment here as well.